### PR TITLE
fix alert system on low message input

### LINKED ISSUE #6770: Alerts not Triggering
We have a 4 node cluster setup and alerts are not triggering unless we have a sustained message  processing of >20 per second.

## Expected Behavior
When an event comes in an alert is triggered as soon as the message comes in.  In our case an email alert when a message contains "Started Network Time Service".

## Current Behavior
When the message comes in we may or may not receive an alert depending on how chatty our systems are being to our Graylog instance.  It can range from a minute later to whenever graylog gets busy enough to send it.  We've had the alert come in weeks later.  Our cluster is up and running but not production.   When setting up the Event Definition we can see the message in the Filter Preview but no alert is ever generated until our graylog server gets busy.  When we enable the Random HTTP Message Generator then the alerts will come through in less than a minute.

## Possible Solution
Fix what ever process logic was added in to version 3.1 to alerts.  We can't have an alert missed in the middle of the night when our systems are less chatty.  That just doesn't make sense for an alert system.  Continuously spamming our Graylog server is not ideal or helpful.

## Steps to Reproduce (for bugs)
1. Setup 4 node cluster
2. Forward a servers's Syslog to Graylog
3. Create an Event Definition and Alert
4. Trigger the Event and watch the Alert is never created
5. Enable the Random HTTP Message Generator and the alert will come through shortly after.

## Context
We need proper alerting for events that come in when they come in.

## Your Environment
4 Node Centos 7.5 Cluster

* Graylog Version: 3.1.3
* Elasticsearch Version: 6.8.3-1
* MongoDB Version: 4.0.12-1
* Operating System:  CentOS Linux release 7.5.1804
* Browser version: Any
[server.txt](https://github.com/Graylog2/graylog2-server/files/3838157/server.txt)

I've attached our server config (from the master)

There are posts related to this [here](https://community.graylog.org/t/alerts-events-do-not-trigger/11792/48) and [here](https://github.com/Graylog2/graylog2-server/issues/6453)

### PR DESCRIPTION
**Pull request by AntonEbel**
## Description

In our current implementation of the alert system, there is a check that is responsible for ensuring that all nodes have indexed their messages by a certain time. Only if this check is successful, we create events. However, there are two scenarios in which the alert system does not create events in time. 

- The first one is when we have multiple nodes and the message rate decreases. In this case we always select the post indexing time of the node with the oldest post indexing time and use it for the previously mentioned check. The danger here is that one node not receiving messages for some time would be enough to prevent the alert system from creating events, since the post-indexing time of that node would always be used at a low message rate.

- The second scenario is when a message arrives that leads to an event, and no other message arrives for some time afterwards. In this case, the event would be created too late. Like in this issue #13228.  The following image illustrates this. 
AlertSystem

This PR fixes the above two cases, but uses a different approach. The general approach is to determine if there are busy nodes in the cluster, and if there are, wait for them to index the messages until a certain time.

close #6770

/jenkins-pr-deps Graylog2/graylog-plugin-enterprise#4194

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog/events/processor/EventProcessorDependencyCheck.java [hunk 1]
```diff
@@ -62,14 +65,10 @@ public boolean canProcessTimerange(DateTime maxTimestamp, Set<String> processorD
...
+     * @param timeRange the timestamp to check
      * @return true if messages up to the given latestTimestamp have already been indexed, false otherwise
      */
-    public boolean hasMessagesIndexedUpTo(DateTime latestTimestamp) {
-        // If there is no timestamp returned, there is no node in the cluster that is processing any messages.
-        // In that case we return false because there is nothing to process.
-        return processingStatusService.earliestPostIndexingTimestamp()
-                .map(latestPostIndexTimestamp -> latestPostIndexTimestamp.isAfter(latestTimestamp) || latestPostIndexTimestamp.isEqual(latestTimestamp))
-                .orElse(false);
+    public boolean hasMessagesIndexedUpTo(TimeRange timeRange) {
```
Comment (mpfz0r) [inline graylog2-server/src/main/java/org/graylog/events/processor/EventProcessorDependencyCheck.java:71]: Change this signature would need another PR for the new SigmaEventProcessor.
But do we need that? (see below)

File: graylog2-server/src/main/java/org/graylog2/system/processing/DBProcessingStatusService.java [hunk 2]
```diff
@@ -195,4 +140,51 @@ ProcessingStatusDto save(ProcessingStatusRecorder processingStatusRecorder, Date
...
+            int behindNodes = 0;
+            while (statusCursor.hasNext()) {
+                activeNodes++;
+                ProcessingStatusDto nodeProcessingStatus = statusCursor.next();
+                DateTime postIndexing = nodeProcessingStatus.receiveTimes().postIndexing();
+                if (postIndexingBehind(timeRange.getTo(), postIndexing)) {
+                    if (isBusy(nodeProcessingStatus)) {
+                        return ProcessingNodesState.SOME_BUSY;
+                    }
+                    if (postIndexing.isBefore(timeRange.getFrom())) {
```
Comment (mpfz0r) [inline graylog2-server/src/main/java/org/graylog2/system/processing/DBProcessingStatusService.java:159]: I don't understand why we are comparing the beginning of the timerange here.
Maybe add some comments to the code?

File: graylog2-server/src/main/java/org/graylog2/system/processing/DBProcessingStatusService.java [hunk 3]
```diff
@@ -195,4 +140,51 @@ ProcessingStatusDto save(ProcessingStatusRecorder processingStatusRecorder, Date
...
+        }
+
+        return ProcessingNodesState.ALL_UP_TO_DATE;
+    }
+
+    private boolean isBusy(ProcessingStatusDto nodeProcessingStatus) {
+        return nodeProcessingStatus.inputJournal().uncommittedEntries() > 0L;
+    }
+
+    private boolean postIndexingBehind(DateTime timeRangeTo, DateTime postIndexing) {
```
Comment (mpfz0r) [inline graylog2-server/src/main/java/org/graylog2/system/processing/DBProcessingStatusService.java:176]: I would just inline this method. I found it easier to understand if I see that `.isAfter` call directly

**Review by mpfz0r — CHANGES_REQUESTED** [review]
First round of review comments

File: graylog2-server/src/main/java/org/graylog/events/processor/EventProcessorDependencyCheck.java (hunk 1 above)
Comment (AntonEbel) [reply graylog2-server/src/main/java/org/graylog/events/processor/EventProcessorDependencyCheck.java:71]: https://github.com/Graylog2/graylog-plugin-enterprise/pull/4194

**Review by mpfz0r — APPROVED** [review]

**Pull request merged**